### ISSUE TITLE
starting mysql - can't wait correct time...

### ISSUE THREAD

**Issue by anztenney** [rhuss/docker-maven-plugin#328]
I'm trying to start a mysql docker before the integration test phase.  Here is the excerpt from my pom.xml:
```xml
<plugins>
    <plugin>
    	<groupId>org.jolokia</groupId>
    	<artifactId>docker-maven-plugin</artifactId>
    	<version>0.13.6</version>
    	<configuration>
    		<images>
    			<image>
    				<alias>database</alias>
    				<name>mysql:5.7</name>
    				<run>
    					<wait>
    						<log>mysqld: ready for connections</log>
    						<time>20000</time>
    					</wait>
    					<env>
    						<MYSQL_ROOT_PASSWORD>abc123</MYSQL_ROOT_PASSWORD>
    						<MYSQL_DATABASE>testdb</MYSQL_DATABASE>
    						<MYSQL_USER>mysql</MYSQL_USER>
    						<MYSQL_PASSWORD>mysql</MYSQL_PASSWORD>
    					</env>
    					<ports>
    						<port>3306:3306</port>
    					</ports>
    				</run>
    			</image>
    		</images>
    	</configuration>
    	<executions>
    		<execution>
    			<id>docker:start</id>
    			<phase>pre-integration-test</phase>
    			<goals>
    				<goal>start</goal>
    			</goals>
    		</execution>
    		<execution>
    			<id>docker:stop</id>
    			<phase>post-integration-test</phase>
    			<goals>
    				<goal>stop</goal>
    			</goals>
    		</execution>
    	</executions>
    </plugin>
    <plugin>
    	<artifactId>maven-failsafe-plugin</artifactId>
    	<version>2.17</version>
    	<executions>
    		<execution>
    			<id>integration-test</id>
    			<goals>
    				<goal>integration-test</goal>
    			</goals>
    		</execution>
    		<execution>
    			<id>verify</id>
    			<goals>
    				<goal>verify</goal>
    			</goals>
    		</execution>
    	</executions>
    	<configuration>
    		<systemPropertyVariables>
    			<mysql.jdbc.url>jdbc:mysql://${docker.host.address}/testdb</mysql.jdbc.url>
    		</systemPropertyVariables>
    	</configuration>
    </plugin>
</plugins>
```

The docker container starts, and my test runs, but it gets a **Caused by: java.net.ConnectException: Connection refused: connect**
```java
conn = DriverManager.getConnection(mysqlUrl, user, password);
stmt = conn.createStatement();
stmt.execute(IOUtils.toString(getClass().getClassLoader().getResource("mysqlScript.sql")));
```

However, if I add a `Thread.sleep(5000)` before I call getConnection, it is able to connect without a problem.  I assume this is some issue with the docker port/container not being immediately available.  Is there any combination in the `<wait>` section to wait an extra 5 seconds, or some other way to make sure this works?

**Comment by jgangemi** [comment]
i do something very similar for my postgres database, although i only wait on a `log` and i have no issues. are you sure mysql is actually up accepting connections by the time `getConnection()` is called?

you could specify a `time` condition on it's own and the plugin would block for that specified time, but i realize that might not be idea. we'd need to add an enhancement to allow blocking for a specified additional interval since `time` cannot be used for this.

**Comment by anztenney** [comment]
I assume it must not be up yet, but I don't know.  I'm waiting for the last line in the log file and then my test runs and fails.  I could do just a `time` condition, but I figured I'd check to see if I was missing some other option since it seems like mysql should be accepting connections.

**Comment by rhuss** [comment]
Waiting for certain log output is here probably really the best solution, however as it seems, MySQL is not really ready immediately after printing out this log entry. The `time` parameter here is indeed a timeout, if `time` is given without any other condition it's a simple wait. The difference is, when used as a timeout, the test fails when it has to wait for that amount of time, for a wait it only blocks. This was discussed in some length in another issue (don't know the number by heart).

On the other hand, whether you wait in the configuration for a certain amount of time or directly within the code, I guess it's not so much of a difference ?

When we have implemented #304 properly then things will get easier because then we can indeed wait on the availability of certain (non-http) port. However there are some technical issues still because of how Docker does the exposure of ports (it opens ports immediately and proxies it through to the image)

**Comment by LeeBreisacher** [comment]
I am running into this using the <log> wait. It waits for the last log entry to be written, but that is apparently not quite enough because tests run immediately after that fail. If I just brute force <time> wait then it works ok (if I wait long enough of course). I tried using <tcp><port>, but that doesn't wait at all -- maybe I'm using the wrong value for <mode>? Can someone give me an exact <wait> that will work for mysql?
